src/tango/quic/crypto/fd_quic_crypto_suites.c

```diff
@@ -811,3 +811,289 @@ fd_quic_crypto_rand( uchar * buf,
+  uchar iv[FD_QUIC_NONCE_SZ] = { 0 };
+
+  /* The AAD is the client IPv4 address, UDP port, and retry source connection id. */
+  ulong aad_sz = (ulong)FD_QUIC_RETRY_TOKEN_AAD_SZ + retry_src_conn_id->sz;
+  uchar aad[aad_sz];
+  memcpy( aad, &ip_addr, sizeof( uint ) );
+  memcpy( aad + sizeof( uint ), &udp_port, sizeof( ushort ) );
+  memcpy( aad + sizeof( uint ) + sizeof( ushort ), &retry_src_conn_id->sz, sizeof( uchar ) );
+  memcpy(
+      aad + sizeof( uint ) + sizeof( ushort ) + sizeof( uchar ),
+      &retry_src_conn_id->conn_id,
+      retry_src_conn_id->sz
+  );
+
+  uchar plaintext[FD_QUIC_RETRY_TOKEN_PLAINTEXT_SZ];
+  memcpy( plaintext, &orig_dst_conn_id->sz, sizeof( uchar ) );
+  memcpy( plaintext + 1, orig_dst_conn_id->conn_id, orig_dst_conn_id->sz );
+  memcpy( plaintext + 1 + orig_dst_conn_id->sz, &now, sizeof( ulong ) );
+
+  /* Append the ciphertext after random bytes in the retry_token. */
+  uchar * ciphertext = hkdf_key + FD_QUIC_RETRY_TOKEN_HKDF_KEY_SZ;
+
+  /* Append the authentication tag after ciphertext in the retry_token. */
+  uchar * tag = ciphertext + FD_QUIC_RETRY_TOKEN_PLAINTEXT_SZ;
+
+  int ciphertext_len = gcm_encrypt(
+      EVP_aes_256_gcm(),
+      plaintext,
+      FD_QUIC_RETRY_TOKEN_PLAINTEXT_SZ,
+      aad,
+      (int)aad_sz,
+      aead_key,
+      iv,
+      ciphertext,
+      tag
+  );
+  if ( FD_QUIC_RETRY_TOKEN_CIPHERTEXT_SZ != ciphertext_len ) {
+    FD_LOG_WARNING(
+        ( "Expected ciphertext length to equal plaintext length. Instead got: %d", ciphertext_len )
+    );
+    return FD_QUIC_FAILED;
+  }
+  return FD_QUIC_SUCCESS;
+}
+
+int fd_quic_retry_token_decrypt(
+    uchar               retry_token[static FD_QUIC_RETRY_TOKEN_CIPHERTEXT_SZ],
+    fd_quic_conn_id_t * retry_src_conn_id,
+    uint                ip_addr,
+    ushort              udp_port,
+    fd_quic_conn_id_t * orig_dst_conn_id,
+    ulong *             now
+) {
+  /* Regenerate the AEAD key (the HKDF key is the first 32 bytes of the token). */
+  uchar * hkdf_key = retry_token;
+  uchar   aead_key[FD_QUIC_RETRY_TOKEN_AEAD_KEY_SZ];
+  fd_quic_hkdf_expand_label(
+      aead_key,
+      FD_QUIC_RETRY_TOKEN_AEAD_KEY_SZ,
+      hkdf_key,
+      FD_QUIC_RETRY_TOKEN_HKDF_KEY_SZ,
+      FD_QUIC_RETRY_TOKEN_AEAD_INFO,
+      FD_QUIC_RETRY_TOKEN_AEAD_INFO_SZ,
+      fd_hmac_sha256,
+      FD_QUIC_RETRY_TOKEN_AEAD_KEY_SZ
+  );
+
+  uchar * ciphertext = hkdf_key + FD_QUIC_RETRY_TOKEN_HKDF_KEY_SZ;
+  ulong   aad_sz     = (ulong)FD_QUIC_RETRY_TOKEN_AAD_SZ + retry_src_conn_id->sz;
+  uchar   aad[aad_sz];
+  memcpy( aad, &ip_addr, sizeof( uint ) );
+  memcpy( aad + sizeof( uint ), &udp_port, sizeof( ushort ) );
+  memcpy( aad + sizeof( uint ) + sizeof( ushort ), &retry_src_conn_id->sz, sizeof( uchar ) );
+  if ( FD_LIKELY( retry_src_conn_id->sz ) ) {
+    fd_memcpy(
+        aad + sizeof( uint ) + sizeof( ushort ) + sizeof( uchar ),
+        &retry_src_conn_id->conn_id,
+        retry_src_conn_id->sz
+    );
+  }
```

Comment:
yeah, the aead stuff is a little duplicative, and also the openssl API usage generally too. i did a pass at refactoring it to use the same logic, but it became messy and larger change than i wanted. the existing aead stuff is a bit coupled with things that only exist in a packet context (hedaers, hp keys, encryption level suites etc.)

i was thinking since we're ripping out openssl crypto soon anyways, we could refactor then to use the same. the retry can be changed to use aes-128 as well